### DISCUSSION TITLE
queryParam - Issue with repeating

### DISCUSSION THREAD

**Discussion by block53** [mockoon/mockoon#1146] [Help]
Hi Guys, This is what I'm trying to do and I'm not sure if it's doable as it is not working for me. Thanks!

Scenario 1 
POST request 
```
{
        "source": "WEB",
        "orderId": "OD01",
        "items": 
         {
            "id": "Item1",
        } 
}
```

RESPONSE 
```
{
    "orderId": "OD01",
    "user": "User1",
    "state": "In_Progress",
    "items": [
		{
			"id": "Item1"
		}
    ]
}
```

Scenario 2 
POST request 
```
{
	"source": "WEB",
	"orderId": "OD01",
   	"items": 
        {
            "id": "Item1",
        }, 
		{
            "id": "Item1",
        } 
}
```

RESPONSE 
```
{
    "orderId": "OD01",
    "user": "User1",
    "state": "In_Progress",
    "items": [
		{
			"id": "Item1"
		},
		{
			"id": "Item1"
		}
    ]
}
```

This is what I have in template but it's not working. I need to repeat id based on the number of id in the request. When returning more than 1 id tag, there will be commas in the number of items so how do I deal with it? 

```
{
    "orderId": "OD01",
    "user": "User1",
    "state": "In_Progress",
    "items": [
		{{#repeat (queryParam 'items.id')}}
        {
      			"id": "{{ body 'items.id' }}",      			
        },
		{{/repeat}}
    ]
}
```

**Comment by 255kb** [comment]
First thing I noticed is that your request JSONs are invalid:

```
{
        "source": "WEB",
        "orderId": "OD01",
        "items":                <-- missing opening bracket
         {
            "id": "Item1",    <-- invalid trailing comma
        }                           <-- missing closing bracket 
}
```

```
{
	"source": "WEB",
	"orderId": "OD01",
   	"items":                 <-- missing opening bracket
        {
            "id": "Item1",     <-- invalid trailing comma
        }, 
		{
            "id": "Item1",     <-- invalid trailing comma
        }                            <-- missing closing bracket 
}
```

When invalid, the entering bodies won't be parsed and you will not be able to use them as objects. 

Aside from this, I would use the [`body`](https://mockoon.com/docs/latest/templating/mockoon-request-helpers/#body) helper in your template: 

```
{
    "orderId": "OD01",
    "user": "User1",
    "state": "In_Progress",
    "items": {{body 'items'}}
}
```

I will automatically get what is at the `items` path and stringify it to be used in JSON text.
If you want to manipulate it before stringifying anything you can use [`bodyRaw`](https://mockoon.com/docs/latest/templating/mockoon-request-helpers/#bodyraw) that can be used with `each` for example.

If you are using a query param to describe which id you want to return, it depends how you are passing this info. If it's a string `item1,item2` you may have to use the [`split`](https://mockoon.com/docs/latest/templating/mockoon-helpers/#split) helper,and then the `filter` and the `bodyRaw` helper. 

Here is a try: 

```
{
    "orderId": "OD01",
    "user": "User1",
    "state": "In_Progress",
    "items": [
    {{#each (split (queryParam 'ids') ',')}}
      {{{stringify (oneOf (filter (bodyRaw 'items') (object id=this)))}}}
      {{#unless @last}},{{/unless}}
    {{/each}}
    ]
}
```

Not saying it's the only solution, but it seems to be working (you may also use `split`, `len` and `repeat`). 
We will soon release v5.0.0 that will also support JSONPath and may make filtering easier, especially this part: `oneOf (filter (bodyRaw 'items') (object id=this))`.
